devicetree: add DT_NODE_FULL_NAME()

galak opened the pull request:
This macro returns a node's name, given its node identifier.

The node name is useful information for the user to utilize for debug
information, similar to DT_NODE_PATH, or DT_LABEL.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
devicetree: add DT_NODE_FULL_NAME()

This macro returns a node's name with unit-adddress, given its node
identifier.

The node name is useful information for the user to utilize for debug
information, similar to DT_NODE_PATH, or DT_LABEL.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

gmarull commented on scripts/dts/gen_defines.py:
```diff
@@ -98,6 +98,9 @@ def main():
             out_comment(f"Node's full path:")
             out_dt_define(f"{node.z_path_id}_PATH", f'"{escape(node.path)}"')
 
+            out_comment(f"Node's name:")
```
```suggestion
            out_comment("Node's name:")
```

galak replied:
@gmarull thanks.  I stole the pattern from a few lines above.

galak committed:
dts: gen_defines: Change f-string to normal string

Change the use of a f-string to a normal string as there is nothing
that needs formatting in the particular instance.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak reviewed (commented):
Should we call this `DT_NODE_FULLNAME`, `DT_NODE_FQN`?

pabigot reviewed (commented):
Copied from #32627 since I hadn't seen this:

As I noted in dev-review my main concern is with precise use of devicetree terminology:

> Each node in the devicetree is named according to the following convention:
>
>     node-name@unit-address
>  The node-name component specifies the name of the node....The unit-address component of the name is specific to the bus type on which the node sits.

I'd like to see the combined identifier called something like `DT_NODE_NAME_AT_UNIT_ADDRESS`.   I really would expect `DT_NODE_NAME` to just be the node-name component.

There is a risk this will not be unique, but if it isn't we can make it the SOC maintainer's responsibility to add a `label` property which will take priority.

However if nobody shares my concern, let it be.

galak commented:
> I'd like to see the combined identifier called something like `DT_NODE_NAME_AT_UNIT_ADDRESS`. I really would expect `DT_NODE_NAME` to just be the node-name component.
> 
> There is a risk this will not be unique, but if it isn't we can make it the SOC maintainer's responsibility to add a `label` property which will take priority.
> 
> However if nobody shares my concern, let it be.

I'm in agreement that the macro should be renamed.  List of possible names:

* `DT_NODE_NAME_AT_UNIT_ADDRESS`
* `DT_NODE_NAME_WITH_UNIT_ADDRESS`
* `DT_NODE_FULLNAME`
* `DT_NODE_FQN`

pabigot commented:
My preference order by weight:

* `DT_NODE_NAME_WITH_UNIT_ADDRESS` 90%
* `DT_NODE_NAME_AT_UNIT_ADDRESS` 8%
* `DT_NODE_FULLNAME` 2%
* `DT_NODE_FQN` nak

erwango commented:
I'm ,not sure what the `UNIT` is referring to. What about `DT_NODE_NAME_WITH_SOC_ADDRESS` ?

pabigot commented:
> I'm ,not sure what the `UNIT` is referring to. What about `DT_NODE_NAME_WITH_SOC_ADDRESS` ?

Nak.  It's nothing necessarily to do with the SoC; the unit address is the first cell of the node's `reg` property (if it has a `reg` property, which is optional).  For some peripheral's it's an address in the SOC address space.  For I2C it's a bus address; for SPI it's a chip-select index.

galak commented:
> Nak. It's nothing necessarily to do with the SoC; the unit address is the first cell of the node's `reg` property (if it has a `reg` property, which is optional). For some peripherals it's an address in the SOC address space. For I2C it's a bus address; for SPI it's a chip-select index.

As @pabigot this unit-address comes from devicetree specification.

![node-name](https://user-images.githubusercontent.com/16417047/109191291-a0e4a380-775b-11eb-8707-4fe6539d60d0.png)

mbolivar-nordic commented:
My two cents in order of preference:

- DT_NODE_FULL_NAME
- DT_NODE_NAME_WITH_UNIT_ADDRESS
- DT_NODE_FULLNAME

jfischer-no commented:
* DT_NODE_FQN
* DT_NODE_FULL_NAME
* DT_NODE_FULLNAME

erwango commented:
`DT_NODE_FULL_NAME` would be nice. Then people can check definition to see what 'full' refers to.

galak commented:
Ok, went with `DT_NODE_FULL_NAME`

pabigot commented on tests/lib/devicetree/api/src/main.c:
```diff
@@ -1625,6 +1625,13 @@ static void test_path(void)
 
+static void test_node_name(void)
+{
+	zassert_true(!strcmp(DT_NODE_FULL_NAME(DT_ROOT), "/"), "");
```
Please add this or something similar, so we verify it works when there is no unit address:
```
	zassert_true(!strcmp(DT_NODE_FULL_NAME(TEST_TEMP),
			     "temperature-sensor"), "");
```

galak replied:
Added

pabigot reviewed (changes requested):
One test improvement please, otherwise this seems fine.

pabigot commented on tests/lib/devicetree/api/src/main.c:
```diff
@@ -1625,6 +1625,17 @@ static void test_path(void)
+	zassert_true(!strcmp(DT_NODE_FULL_NAME(TEST_TEMP),
+			     "temperature-sensor"), "");
+	zassert_true(strcmp(DT_NODE_FULL_NAME(TEST_REG),
+			     "reg-holder"), "");
```
Why isn't this `"reg-holder@9999aaaa"`?

galak replied:
Because I'm testing that it does NOT match.

pabigot replied:
Ah.  A little subtle.
